anthony/unrealized-potential-of-string-diagrams.md

```diff
@@ -0,0 +1,128 @@
+---
+title: The Unrealized Potential of String Diagrams for ZKVMs
+publish_date: ???
+category: compilers
+image: media/string_puppet.webp
+excerpt: This post will describe string diagrams and explain why they are a well suited for designing ZKVMs.
+---
+
+## Trees Considered Harmful
+
+By the end of this post, I want to convince the reader that there are better forms of syntax than abstract syntax trees; namely, graphical ones. I'll also touch on how this may influence better ZKVM designs.
+
+First, *what's so bad about trees?* 
+
+Historically, most languages in mathematics have been constructed using abstract syntax trees, or ASTs. The very definition of a grammar in most computer science contexts assumes that any term of any language must be structured like a tree. Most of the time, this seems fine, but a century of hacks have had to be invented to compensate for the deficiencies of trees.
+
+I think the easiest place to see this is in the combinator calculus. Variables remain a persistant issue for theory, and much ink has been spilt trying to treat variables, or really, variable names, coherently. Issues related to scoping and shadowing formally complicate the syntactical aspects of mathematics which most simply take for granted. Among the earliest methods for managing them is to simply eliminate them. From this, we get the original combinator calculus invented by Skonfinkle. Conceptually similar systems were invented over the last century for similar reasons, such as [predicate functor logic](https://en.wikipedia.org/wiki/Predicate_functor_logic). These allow one to reason through symbol manipulation without needing named variables, but one must spend an inordinate effort dealing with a convoluted origami just to get separate ends of trees to communicate.
+
+Another approach to eliminating the variables created by quantifiers in first-order logic is [Relation algebra](https://en.wikipedia.org/wiki/Relation_algebra). This system is extremely interesting, and has been developed these days largely under the guise of the algebra of programming.
+
+- [Components and acyclicity of graphs. An exercise in combining precision with concision](https://www.sciencedirect.com/science/article/pii/S2352220821000936)
+
+however, it has some obvious downsides. The core relation algebra is not as expressive as the full first-order logic. It's equivalent to the first-order logic when restricted to three variables. One can midigate this issue by manually introducing a notion of tuple, along with some relations describing tuple manipulation (forking, projecting, etc.). This puts one in a similar place to combinator logics where one is forced to deal with an awkward yoga of point-free operators to do things which would be taken for granted in a setting with variable binding.
```

Comment:
another big knowledge nuke

can we appease the reader with some nice pictures stolen, e.g., from here ?

https://www.cas.mcmaster.ca/~kahl/Publications/Conf/Kahl-2001e.pdf

Reply:
Hmm... I haven't seen this before.